File: 202.Happy-Number.js
/**
 * @param {number} n
 * @return {boolean}
 */
const digitSum = (n) => {
  let sum = 0;

  while (n) {
    const rem = n % 10;
    n = Math.floor(n / 10);
    sum += rem * rem;
  }

  return sum;
};

var isHappy = function (n) {
  let slow = n;
  let fast = n;

  while (true) {
    slow = digitSum(slow);
    fast = digitSum(digitSum(fast));

    if (slow === fast) {
      return slow === 1;
    }
  }

  return false;
};

File: python/202.Happy-Number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        slow = n
        fast = n

        def digitSum(num):
            total = 0

            while num:
                rem = num % 10
                total += (rem * rem)
                num = num // 10

            return total

        while (True):
            slow = digitSum(slow)
            fast = digitSum(digitSum(fast))

            if slow == fast:
                return slow == 1
        
        return False